“Improve Sensorless homing/probing accuracy for G28, G33, M48”

lujios opened the pull request:
### Requirements
* Trigorilla V1 ATmega2560
* Drivers TMC2209
* Enabled SENSORLESS_HOMING
* Enabled SENSORLESS_PROBING
* Enabled IMPROVE_HOMING_RELIABILITY


### Description
It improves the features introduced with “Improve Sensorless homing/probing for G28, G33 #21899”.
Changes:
1.	**G28.cpp**
To get the TMCs to stabilize when they are intervened (change the operating current or deactivate/activate STALLGUARD), a delay is included.
`safe_delay(SENSORLESS_STALLGUARD_DELAY); // Short delay needed to settle`
It depends on `SENSORLESS_STALLGUARD_DELAY`, with a default value of 0 ms if it was not declared. I declared it with a value of 200 ms in Configuration.h.

2.	**G33.cpp**
I add the modifier 'S' `save_offset_sensorless_adj`, which allows G33 to save an offset on the motor whose STALLGUARD mode is disabled. It can be used in conjunction with the X, Y, or Z modifiers in pairs, so that STALLGUARD will only be disabled on a single engine. Or only with levels higher than V0, so that it automatically loads `offset_sensorless_adj` at the start of the polling cycle.
The `set_homing_current()` function is now declared in 'endstop.h'. That is why it is invoked on the 'endstops.' class. And it is made to depend again on SENSORLESS_HOMING (change made in "G33 uses a probe, even when a probe is not defined #22909" to avoid the compilation error when SENSORLESS_PROBING was not used, an error that was due to the fact that the library "probe.h" depended on HAS_BED_PROBE)
To achieve a probe radius less than `DELTA_PRINTABLE_RADIUS`, a new variable is declared, which multiplied over the variable “delta calibration radius” `dcr` allows the dcr value to be reduced. I have verified that better results are obtained when the probe is run below 70% `dcr`.
To automatically save the `offset_sensorless_adj` corresponding to each motor and to only run with test levels higher than V0 and when `save_offset_sensorless_adj` is active, 3 individual tests are added to the center of the bed, where STALLGUARD will be deactivated independently in each TMC. The goal is to load `offset_sensorless_adj`, with the new `probe.save_offset_sensorless()` function.
For test levels V0 and P1 in the case of having `HAS_DELTA_SENSORLESS_PROBING` active, `probe.save_offset_sensorless()` can be used to load the value of `z_at_pt[CEN]` in the motors where STALLGUARD TMC is deactivated.
At the end of the G33 code, `probe.test_sensitivity` is activated in all three axes. So that it does not interfere with the rest of the probing test (M48, G29, etc), in the case of having used the X, Y or Z modifiers with G33.


3.	**Conditionals_adv.h**
We validate the activation of the use with SENSORLESS (`USE_SENSORLESS`), to define the `SENSORLESS_STALLGUARD_DELAY` with a value of 0, in case it has not been declared.

4.	**menu_delta_calibrate.cpp**
I add a modifiable value to apply a reduction factor in “delta calibration radius” or `dcr`.
The calculation of “delta calibration radius” or `dcr`, which was unlinked from `PROBING_MARGIN` and `offset_xy`, is corrected. And I add the reduction factor `sensorless_radius_factor` that is applied on dcr when `HAS_DELTA_SENSORLESS_PROBING` is defined.

5.	**delta.cpp**
Variables are declared;
`offset_sensorless_adj{0}`. Vector that stores the setting applicable to the driver that activates the endstop during probing.
`offset_sensorless`. Variable applicable to the offset of each test point during probing.
`sensorless_radius_factor`. Reduction factor dcr for the probing test.
I add;
`safe_delay(SENSORLESS_STALLGUARD_DELAY)` for TMCs to stabilize when disabling/enabling stealthChop mode and enabling/disabling driver pin diag1.

6.	**delta.h**
External variables are declared;
`offset_sensorless_adj`. Vector that stores the setting applicable to the driver that activates the endstop during probing.
`offset_sensorless`. Variable applicable to the offset of each test point during probing.
`sensorless_radius_factor`. Reduction factor dcr for delta probing test.
7.	endstops.cpp y endstops.h
Move here the `set_homing_current()` function that was in probe.cpp. It is done to declare it in a file common to the use of G28 and G33 and that `SENSORLESS_HOMING` and `SENSORLESS_PROBING` are independent.
The library for debugging, which needs the function, is added.
`#include "../core/debug_out.h"`

8.	**motion.cpp**
I add;
`safe_delay(SENSORLESS_STALLGUARD_DELAY)` for TMCs to stabilize on disable/enable stealthChop mode and enable/disable driver pin diag1 if `SENSORLESS_HOMING` is active.
9.	probe.cpp y probe.h
By default, `test_sensitivity = {true, true, true}` is activated.
`save_offset_sensorless` is evaluated to apply the corresponding sensorless offset after each probe detected.
We apply `offset_sensorless` to each `probe_z` value that will be used in the autoprobe calculations.
At the end of each probed point, a message of the value in `measured_z` is displayed to be activated with DEBUG.
The new function `save_offset_sensorless()` is declared and we move `set_homing_current()` to endstops.cpp/h.


<!--

We must be able to understand your proposed change from this description. If we can't understand what the code will do from this description, the Pull Request may be closed at the maintainers' discretion. Keep in mind that the maintainer reviewing this PR may not be familiar with or have worked with the code recently, so please walk us through the concepts.

-->

### Benefits
Improves G33 autocalibration accuracy, when `USE_SENSORLESS` is enabled on deltas.

<!-- What does this fix or improve? -->

### Configurations
[Configuration.zip](https://github.com/MarlinFirmware/Marlin/files/8747506/Configuration.zip)

<!-- Attach any Configuration.h, Configuration_adv.h, or platformio.ini files needed to compile/test your Pull Request. -->

### Related Issues
#21899
#22909
<!-- Whether this fixes a bug or fulfills a feature request, please list any related Issues here. -->

lujios committed:
Mejora SENSORLESS en DELTA

Cambios en G28 y G33 para hacer funcional SENSORLESS_HOMING y SENSORLESS_PROBING. Se incluyen nuevas funciones en probe.cpp

lujios committed:
Merge branch 'MarlinFirmware:bugfix-2.0.x' into bugfix-2.0.x

lujios committed:
Update probe.cpp

lujios committed:
Error versión actualizada

Regreso a una versión anterior

lujios committed:
Error versión actualizada

Regreso a una versión anterior

lujios committed:
DELTA with SENSORLESS_HOMING/PROBING.

Enhancements for optimal G28 and G33 performance in DELTA with SENSORLESS_PROBING and SENSORLESS_HOMING enabled.

lujios committed:
Merge branch 'MarlinFirmware:bugfix-2.0.x' into bugfix-2.0.x

lujios committed:
Change description.

Change description Probe::current_homing_on().

lujios committed:
Revert "Change description."

This reverts commit 22671e0c0567d241e715523eb94e6ea5dc38704b.

lujios committed:
Revert "Revert "Change description.""

This reverts commit e842e7ad3a32cd1ada42b1a5f21a750a8d12ca25.

lujios committed:
Delete Configuration.h

lujios committed:
Delete Configuration_adv.h

lujios committed:
Configuration.

Copy Configuration.h and Configuration_adv.h

lujios committed:
Merge branch 'bugfix-2.0.x' of https://github.com/lujios/Marlin into bugfix-2.0.x

lujios committed:
Merge branch 'MarlinFirmware:bugfix-2.0.x' into bugfix-2.0.x

lujios committed:
Change Configuration origin

Change Configuration origin

lujios committed:
Merge branch 'bugfix-2.0.x' of https://github.com/lujios/Marlin into bugfix-2.0.x

lujios committed:
Update pins_RAMPS.h

Changed pins_RAMPS.h to origin.

lujios committed:
Update G28.cpp

Resolving conflicts between lujios:bugfix-2.0.x and MarlinFirmware:bugfix-2.0.x and committing changes  lujios:bugfix-2.0.x

lujios committed:
Merge branch 'bugfix-2.0.x' into bugfix-2.0.x

lujios committed:
Update G28.cpp

Error #endif without #if. Fix.

lujios committed:
Merge branch 'bugfix-2.0.x' into bugfix-2.0.x

lujios committed:
Merge branch 'MarlinFirmware:bugfix-2.0.x' into bugfix-2.0.x

lujios committed:
Update probe.cpp

"SLOW Probe fail! - No trigger.".
Version 2.0.9, mistakenly modified to see if the probe was triggered, on the MAX endstop, instead of the MIN.

lujios committed:
Update probe.cpp


Correct variable name "save_current_" in 'void Probe :: current_homing_off ()'.

lujios committed:
Update probe.cpp

Oops !! Typographic error.

lujios committed:
Update pins_AZSMZ_MINI.h

X_MIN_PIN and Y_MIN_PIN not defined and a compilation error occurred in "probe.cpp".
Test.

thinkyhead committed:
initial review quick fixes

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/21899

thinkyhead committed:
non-delta slow homing

thinkyhead committed:
fewer static buffers

thinkyhead committed:
style tweaks

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/21899

thinkyhead committed:
Update set_homing_current

thinkyhead committed:
Always trigger MAX on DELTA

thinkyhead committed:
move some to planner

lujios committed:
Update planner.cpp

Error saved_motion_state

lujios committed:
Typographic error. debug_current -> debug_current_on

debug_current -> debug_current_on

lujios committed:
SENSORLESS_PROBING to SENSORLESS_HOMING

ac_home()

lujios committed:
ENABLED(SENSORLESS_PROBING) or ENABLED(SENSORLESS_HOMING)

lujios committed:
ENABLED(SENSORLESS_PROBING) or ENABLED(SENSORLESS_HOMING)

lujios committed:
Update probe.h

lujios committed:
Update probe.cpp

lujios committed:
Update probe.cpp

lujios committed:
Update probe.cpp

lujios committed:
Update probe.cpp

lujios committed:
Update probe.cpp

I must be very clumsy.

lujios committed:
Update probe.cpp

Unable to understand compilation error, with code that is identical elsewhere.
Help!!

lujios committed:
Update probe.cpp

I try to avoid the error during check commit.

lujios committed:
Update probe.cpp

This error is driving me crazy.

lujios committed:
Update probe.cpp

lujios committed:
Update probe.cpp

#include "stepper/trinamic.h"

lujios committed:
Update probe.cpp

lujios committed:
Update tmc_util.h

To resolve conflict in the source branch.
This code had been brought into planner.h for convenience, but I don't think it's necessary. If modified to support DELTA.

lujios committed:
Update probe.h

lujios committed:
Update probe.cpp

New functionality for G33. It does not activate stallguard on the selected axis.

lujios committed:
Update G33.cpp

New functionality that allows disabling stallguard on the selected axis. Allows you to calibrate the sensitivity individually, when using SENSORLESS_PROBING.

lujios committed:
Update planner.h

We move the definition of motion_state_t, to tmc_util.h, to avoid conflicts in the source branch.

lujios committed:
Update endstops.cpp

#if BOTH(DELTA, SENSORLESS_PROBING)

lujios committed:
Update delta.cpp

endstops.report_states() at the end of the first approach in homing DELTA.

lujios committed:
Update probe.cpp

lujios committed:
Update tmc_util.h

lujios committed:
Update tmc_util.h

lujios committed:
Update tmc_util.h

lujios committed:
Update tmc_util.h

lujios committed:
Update G33.cpp

lujios committed:
Update planner.h

lujios committed:
Update planner.cpp

lujios committed:
Update G28.cpp

lujios committed:
Merge branch 'MarlinFirmware:bugfix-2.0.x' into bugfix-2.0.x

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/21899

thinkyhead committed:
Use bitfields for test_sensitivity

thinkyhead committed:
motion_state_t

thinkyhead committed:
Tweak some SPI_ENDSTOPS code

thinkyhead committed:
Fix warning

lujios committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into bugfix-2.0.x

lujios committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into bugfix-2.0.x

lujios committed:
Bug G28.cpp

Duplicate;
#if HAS_CURRENT_HOME(I)
#if HAS_CURRENT_HOME(J)
#if HAS_CURRENT_HOME(K)

lujios committed:
Add files via upload

lujios committed:
Add files via upload

lujios committed:
Add files via upload

lujios committed:
Merge branch 'MarlinFirmware:bugfix-2.0.x' into bugfix-2.0.x

lujios committed:
Merge branch 'MarlinFirmware:bugfix-2.0.x' into bugfix-2.0.x

lujios committed:
Merge branch 'MarlinFirmware:bugfix-2.0.x' into bugfix-2.0.x

lujios committed:
Merge branch 'MarlinFirmware:bugfix-2.0.x' into bugfix-2.0.x

lujios committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into bugfix-2.0.x

lujios committed:
Merge branch 'MarlinFirmware:bugfix-2.0.x' into bugfix-2.0.x

lujios committed:
Limpiar código

lujios committed:
Update

lujios committed:
Add files via upload

lujios committed:
Merge branch 'MarlinFirmware:bugfix-2.0.x' into bugfix-2.0.x

lujios committed:
Mejora SENSORLESS_PROBING

lujios committed:
Add files via upload

lujios committed:
Add files via upload

lujios committed:
Update .gitignore

thinkyhead committed:
🧑‍💻 Strip #errors in Configurations deployment

lujios committed:
Merge branch 'MarlinFirmware:bugfix-2.0.x' into bugfix-2.0.x

lujios committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into bugfix-2.0.x

lujios committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into bugfix-2.0.x

lujios committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into bugfix-2.0.x

lujios committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into bugfix-2.0.x

lujios committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into bugfix-2.0.x

lujios committed:
Update endstops.cpp

debug_current_on

lujios committed:
Update Conditionals_adv.h

[Test FYSETC_F6] RAMPS | Mixed TMC | Sensorless | RRDFGSC | Games...
Marlin/src/module/motion.cpp:1676:20: error: 'SENSORLESS_STALLGUARD_DELAY' was not declared in this scope
         safe_delay(SENSORLESS_STALLGUARD_DELAY); // Short delay needed to settle

lujios committed:
Update G33.cpp

[Test FYSETC_F6] RAMPS 1.3 | DELTA | FLSUN AC Config...
Marlin/src/gcode/calibrate/G33.cpp: In static member function 'static void GcodeSuite::G33()':
Marlin/src/gcode/calibrate/G33.cpp:697:9: error: 'class Probe' has no member named 'test_sensitivity'
   probe.test_sensitivity = {true, true, true};

lujios committed:
Update probe.cpp

[Test FYSETC_F6] FYSETC_F6 | SCARA | Mixed TMC | EEPROM...
Marlin/src/module/probe.cpp:919:5: error: 'offset_sensorless' was not declared in this scope
Marlin/src/module/probe.cpp:921:11: error: 'test_sensitivity' was not declared in this scope

thinkyhead committed:
cleanup

thinkyhead committed:
Better delta sensorless homing/probing accuracy

Co-Authored-By: lujios <83166168+lujios@users.noreply.github.com>

thinkyhead commented:
You and @ellensp will need to hash out the correct flags to use here. I think `SENSORLESS_HOMING` is more correct.

#21899…
```cpp
  TERN_(SENSORLESS_HOMING, probe.set_homing_current(true));
  home_delta();
  TERN_(SENSORLESS_HOMING, probe.set_homing_current(false));
```

#22909…
```cpp
  TERN_(HAS_DELTA_SENSORLESS_PROBING, probe.set_homing_current(true));
  home_delta();
  TERN_(HAS_DELTA_SENSORLESS_PROBING, probe.set_homing_current(false));
```

This PR…
```cpp
  TERN_(SENSORLESS_HOMING, endstops.set_homing_current(true));
  home_delta();
  TERN_(SENSORLESS_HOMING, endstops.set_homing_current(false));
```

lujios commented:
> 2\. 'endstop.h'

In fact, the correct thing to do is to use SENSORRLESS_HOMING. The `ac_home()` function is used whether or not `SENSORLESS_PROBING` is enabled. Associating it with enabling `SENSORLESS_PROBING` for delta would cause abnormal behavior for those configurations where only `SENSORLESS_HOMING` is desired.
As I already mentioned, the error that occurred when compiling and that was fixed in #22909, was not due to an incorrect use of the `SENSORLESS_HOMING` flag, but rather because the `probe.h` library was not declared for the configuration used. And it should be fixed by migrating `set_homing_current()` to `endstop.h`.

lujios commented:
> You and @ellensp will need to hash out the correct flags to use here. I think `SENSORLESS_HOMING` is more correct.
> 
> #21899…
> 
> ```c++
>   TERN_(SENSORLESS_HOMING, probe.set_homing_current(true));
>   home_delta();
>   TERN_(SENSORLESS_HOMING, probe.set_homing_current(false));
> ```
> 
> #22909…
> 
> ```c++
>   TERN_(HAS_DELTA_SENSORLESS_PROBING, probe.set_homing_current(true));
>   home_delta();
>   TERN_(HAS_DELTA_SENSORLESS_PROBING, probe.set_homing_current(false));
> ```
> 
> This PR…
> 
> ```c++
>   TERN_(SENSORLESS_HOMING, endstops.set_homing_current(true));
>   home_delta();
>   TERN_(SENSORLESS_HOMING, endstops.set_homing_current(false));
> ```

lujios commented:
> You and @ellensp will need to hash out the correct flags to use here. I think `SENSORLESS_HOMING` is more correct.
> 
> #21899…
> 
> ```c++
>   TERN_(SENSORLESS_HOMING, probe.set_homing_current(true));
>   home_delta();
>   TERN_(SENSORLESS_HOMING, probe.set_homing_current(false));
> ```
> 
> #22909…
> 
> ```c++
>   TERN_(HAS_DELTA_SENSORLESS_PROBING, probe.set_homing_current(true));
>   home_delta();
>   TERN_(HAS_DELTA_SENSORLESS_PROBING, probe.set_homing_current(false));
> ```
> 
> This PR…
> 
> ```c++
>   TERN_(SENSORLESS_HOMING, endstops.set_homing_current(true));
>   home_delta();
>   TERN_(SENSORLESS_HOMING, endstops.set_homing_current(false));
> ```

In fact, the correct thing to do is to use `SENSORRLESS_HOMING`. The `ac_home()` function is used whether or not `SENSORLESS_PROBING` is enabled. Associating it with enabling `SENSORLESS_PROBING` for delta would cause abnormal behavior for those configurations where only `SENSORLESS_HOMING` is desired.
As I already mentioned, the error that occurred when compiling and that was fixed in #22909, was not due to an incorrect use of the `SENSORLESS_HOMING` flag, but rather because the `probe.h` library was not declared for the configuration used. And it should be fixed by migrating `set_homing_current()` to `endstop.h`.

ellensp commented:
@thinkyhead you changed  SENSORLESS_HOMING to HAS_DELTA_SENSORLESS_PROBING in my PR... 
https://github.com/MarlinFirmware/Marlin/pull/22909/commits/08240f03358226d44a70b6c3ba4d5b77fa02768b

robbycandra committed:
🩹 Adjust manage_media for slow/late media init (USB FD) (#24015)

ellensp committed:
🚸 Firmware upload destination prompt (using Tk) (#24074)

thinkyhead commented:
@ellensp — That is equivalent to `SENSORLESS_PROBING` when `DELTA` is enabled. The tweak was a followup to a6940ec2.

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/24220

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/24220

filippor commented:
> @ellensp — That is equivalent to `SENSORLESS_PROBING` when `DELTA` is enabled. The tweak was a followup to [a6940ec](https://github.com/MarlinFirmware/Marlin/commit/a6940ec2ca9c26757a4ec70707fa4c1dfbeecd1e).

For DELTA  PROBING is different from HOMING
HOMING is on top to reset the position of single axis
PROBING is down on bed

lujios commented:
I think it caused confusion when initially this piece of code, `set_homing_current()`, was housed in the "probe." class. 
As @filippor indicates, in a delta HOMING and PROBING are independent, both in the position where they are activated and the endstops used.
The purpose of the `ac_home()` function is to execute each HOMING required by each PROBING cycle (G33). But being independent, SENSORLESS may have been activated for PROBING and not for HOMING, vice versa or both, but always independent. So in `ac_home()`, `set_homing_current()` will always depend on `SENSORLESS_HOMING`.

thinkyhead committed:
use the cond

filippor commented on Marlin/src/gcode/calibrate/G33.cpp:
```diff
@@ -475,6 +479,27 @@ void GcodeSuite::G33() {
+      probe.test_sensitivity = { true, false, false };
+      if (probe_calibration_points(z_at_pt, 1, dcr, false, false, probe_at_offset))
+        probe.save_offset_sensorless(true, z_at_pt[CEN]);
+      z_at_pt[NPP + 1] = { 0.0f };
```
warning z_at_pt[NPP + 1]   z_at_pt has size NPP + 1
and I think z_at_pt is resetted inside probe_calibration_points

lujios replied:
Perhaps it is not the most orthodox, I do not argue.
The purpose of initializing `z_at_pt[NPP + 1]` is so that `z_at_pt[CEN]` is not saved to each of the three center probes needed to load `offset_sensorless_adj()` (one center probe per axis), none of the three points can be part of deviation statistics for G33.
In fact, I copied line 505, which does the exact same thing at the start of every G33 loop.

filippor replied:
I'm not a c expert but the compiler give me this warnng
Marlin\src\gcode\calibrate\G33.cpp: In static member function 'static void GcodeSuite::G33()':
Marlin\src\gcode\calibrate\G33.cpp:491:22: warning: array subscript 7 is above array bounds of 'float [7]' [-Warray-bounds]
  491 |       z_at_pt[NPP + 1] = { 0.0f };
      |       ~~~~~~~~~~~~~~~^
Marlin\src\gcode\calibrate\G33.cpp:487:13: note: while referencing 'z_at_pt'
  487 |       float z_at_pt[NPP + 1] = { 0.0f };
      |             ^~~~~~~

If I understand correctly with z_at_pt[NPP + 1] = { 0.0f }    you are writing something on index NPP + 1 and the max index  is NPP

filippor replied:
inside  probe_calibration_points   
there is this macro that I think set all value to 0
LOOP_CAL_ALL(rad) z_pt[rad] = 0.0f;

lujios replied:
I don't get the prompt during my builds.
And the same error does not appear in line 505?
There is similar code on line 342, 331. I also don't see any warning during compilation.
Any idea how to fix it, maybe if I declare the float variable every time it is initialized?

lujios replied:
> inside probe_calibration_points there is this macro that I think set all value to 0 LOOP_CAL_ALL(rad) z_pt[rad] = 0.0f;

Thanks, very good suggestion. I fix it like this.

filippor replied:
Line 505 is a new variable declaration not a assignment

lujios committed:
Merge branch 'bugfix-2.0.x' of https://github.com/lujios/Marlin into bugfix-2.0.x

lujios committed:
probe.cpp

Initialize "offset_sensorless" to "save_offset_sensorless(false,).

lujios committed:
Update G33.cpp

Initialize z_at_pt[] whith macro LOOP_CAL_ALL(rad).

lujios committed:
Instability in tests.

The tests I have done, with the changes in the code, have shown instability. I have fixed parts of the code that were targeted for cleanup and responsible for the unexpected behavior.

lujios committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into bugfix-2.0.x

lujios committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into bugfix-2.0.x

lujios committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into bugfix-2.0.x

lujios committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into bugfix-2.0.x

thinkyhead committed:
clean up

thinkyhead committed:
split up set versus refresh

thinkyhead committed:
try a lambda

rondlh commented:
This needs a section in configuration_adv.h otherwise it will never be used...

thisiskeithb commented:
> This needs a section in configuration_adv.h otherwise it will never be used...

And as reported in our Discord, it'd also fix the issue when compiling with the [default Ender-3 with SKR Mini E3 V3 config](https://github.com/MarlinFirmware/Configurations/tree/bugfix-2.0.x/config/examples/Creality/Ender-3%20Pro/BigTreeTech%20SKR%20Mini%20E3%203.0):

```prolog
Compiling .pio/build/STM32G0B1RE_btt/src/src/gcode/control/M993_M994.cpp.o
Marlin/src/gcode/calibrate/G28.cpp: In static member function 'static void GcodeSuite::G28()':
Marlin/src/gcode/calibrate/G28.cpp:324:16: error: 'SENSORLESS_STALLGUARD_DELAY' was not declared in this scope
  324 |     safe_delay(SENSORLESS_STALLGUARD_DELAY); // Short delay needed to settle
      |                ^~~~~~~~~~~~~~~~~~~~~~~~~~~
Compiling .pio/build/STM32G0B1RE_btt/src/src/gcode/control/M997.cpp.o
Compiling .pio/build/STM32G0B1RE_btt/src/src/gcode/control/M999.cpp.o
*** [.pio/build/STM32G0B1RE_btt/src/src/gcode/calibrate/G28.cpp.o] Error 1
=================================== [FAILED] Took 30.75 seconds ===================================

Environment      Status    Duration
---------------  --------  ------------
STM32G0B1RE_btt  FAILED    00:00:30.751
```

lujios commented:
> configuration_adv.h

If I agree, it should be included in configuration_adv.h.
I suppose that the value it takes will depend on the architecture of the printer, as well as its construction. That's why I didn't see him as adequate to face him in this PR.
During my tests, I noticed that if I turned on DEBURING, the triggered false positives were greatly reduced or even disappeared. For this reason, I tried to include a delay every time there were changes to the TMC's configuration. But if true, your sensitivity remains the same. If your settings include values ​​with very high sensitivity, false positives will still occur.
I did my tests in a delta, where all three axes are moved with identical motors and the same or very similar TMC settings. Consider this premise.

lujios commented:
Build fails with "HAS_DELTA_SENSORLESS_PROBING" enabled.
```
Compiling .pio\build\mega2560\src\src\gcode\config\M221.cpp.o
Marlin\src\gcode\calibrate\G33.cpp: In static member function 'static void GcodeSuite::G33()':
Marlin\src\gcode\calibrate\G33.cpp:440:28: error: 'struct Probe::sense_bool_t' has no member named 'set'
     probe.test_sensitivity.set(!parser.seen_test('X'), !parser.seen_test('Y'), !parser.seen_test('Z'));
                            ^~~
Compiling .pio\build\mega2560\src\src\gcode\config\M301.cpp.o
*** [.pio\build\mega2560\src\src\gcode\calibrate\G33.cpp.o] Error 1
============================================================================= [FAILED] Took 31.62 seconds =============================================================================

Environment    Status    Duration
-------------  --------  ------------
mega2560       FAILED    00:00:31.616
======================================================================== 1 failed, 0 succeeded in 00:00:31.616 ========================================================================
```

I have modified the code like this, and the error is resolved;
```
  #if HAS_DELTA_SENSORLESS_PROBING
    probe.test_sensitivity = {!parser.seen_test('X'), !parser.seen_test('Y'), !parser.seen_test('Z')};
    const bool do_save_offset_adj = parser.seen_test('S');
  #endif

```
@thinkyhead  is it possible to include this fix in your branch [fcef8d9](https://github.com/MarlinFirmware/Marlin/commit/fcef8d946cbd45e27ba9888f1e18327b68d52b00)?

thisiskeithb commented:
@lujios: We’ll need another PR.

lujios commented:
> @lujios: We’ll need another PR.

I wanted to say in the same PR where I think @thinkyhead has already corrected the rest of the problems derived from this merger.

thisiskeithb commented:
> I wanted to say in the same PR where I think @thinkyhead has already corrected the rest of the problems derived from this merger.

There's no other PR. The fix was committed directly to the `bugfix-2.0.x` branch: https://github.com/MarlinFirmware/Marlin/commit/3522d5376cbf03658b71ddaf8c341e35f59b369b
